### PR TITLE
RBAC - REST APIs

### PR DESCRIPTION
**Pull request by TLeoDev**
## Description

What's new?

- Creation of GetAll and GetDetails api for Group, User, AccessControl and Role classes.

## What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Tests
- [ ] Other

### CODE REVIEW

File: src/IoTHub.Portal.Shared/Models/v1.0/RoleDetailsModel.cs [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace IoTHub.Portal.Shared.Models.v10
+{
+    using System.Collections.Generic;
+    public class RoleDetailsModel
+    {
+        public string Id { get; set; }
+        public string Name { get; set; }
+        public ICollection<ActionModel> Actions { get; set; } = new List<ActionModel>();
```
Comment (kbeaugrand) [inline src/IoTHub.Portal.Shared/Models/v1.0/RoleDetailsModel.cs:11]: Actions have to be only represented by their name. 
```suggestion
        public ICollection<string> Actions { get; set; } = new List<string>();
```

File: src/IoTHub.Portal.Shared/Models/v1.0/ActionModel.cs [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace IoTHub.Portal.Shared.Models.v10
+{
+    public class ActionModel
+    {
+        public string Id { get; set; }
+        public string Name { get; set; }
+        public string RoleId { get; set; }
+    }
```
Comment (kbeaugrand) [inline src/IoTHub.Portal.Shared/Models/v1.0/ActionModel.cs:11]: This class is useless

File: src/IoTHub.Portal.Shared/Models/v1.0/UserDetailsModel.cs [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+namespace IoTHub.Portal.Shared.Models.v10
+{
+    //using System.Collections.Generic;
+    public class UserDetailsModel
+    {
+        public string Id { get; set; }
+        public string Email { get; set; }
+        public string Name { get; set; }
+        public string Forename { get; set; }
+        //public ICollection<AccessControlModel> AccessControls { get; set; } = new List<AccessControlModel>();
```
Comment (kbeaugrand) [inline src/IoTHub.Portal.Shared/Models/v1.0/UserDetailsModel.cs:13]: Remove useless code

```suggestion
        //public ICollection<AccessControlModel> AccessControls { get; set; } = new List<AccessControlModel>();
```

File: src/IoTHub.Portal.Shared/Models/v1.0/UserModel.cs [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+namespace IoTHub.Portal.Shared.Models.v10
+{
+
+    public class UserModel
+    {
+        public string Id { get; set; }
+        public string Email { get; set; }
+        public string Name { get; set; }
+        public string Forename { get; set; }
+    }
```
Comment (kbeaugrand) [inline src/IoTHub.Portal.Shared/Models/v1.0/UserModel.cs:13]: What is the difference with the UserDetails?

File: src/IoTHub.Portal.Shared/Models/v1.0/GroupModel.cs [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace IoTHub.Portal.Shared.Models.v10
+{
+
+    public class GroupModel
+    {
+        public string Id { get; set; }
+        public string Name { get; set; }
+        public string Avatar { get; set; }
```
Comment (kbeaugrand) [inline src/IoTHub.Portal.Shared/Models/v1.0/GroupModel.cs:11]: Remove the avatar.
```suggestion

```

File: src/IoTHub.Portal.Shared/Models/v1.0/UserModel.cs [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+
+namespace IoTHub.Portal.Shared.Models.v10
+{
+
+    public class UserModel
+    {
+        public string Id { get; set; }
+        public string Email { get; set; }
+        public string Name { get; set; }
+        public string Forename { get; set; }
```
Comment (kbeaugrand) [inline src/IoTHub.Portal.Shared/Models/v1.0/UserModel.cs:12]: Please use these properties in the model.
Think also to adapt the entity and table....

```suggestion
        public string Id { get; set; }
        public string Email { get; set; }
        public string Name { get; set; }
        public string GivenName { get; set; }
        public string FamilyName{ get; set; }
```

**Review by kbeaugrand — CHANGES_REQUESTED** [review]

File: src/IoTHub.Portal.Shared/Models/v1.0/UserDetailsModel.cs (hunk 3 above)
Comment (TLeoDev) [reply src/IoTHub.Portal.Shared/Models/v1.0/UserDetailsModel.cs:13]: This is not useless, it prepares the 'User' class to collect the accessControl list, which has not yet been implemented. It's why i decide to comment it.

File: src/IoTHub.Portal.Shared/Models/v1.0/UserModel.cs (hunk 4 above)
Comment (TLeoDev) [reply src/IoTHub.Portal.Shared/Models/v1.0/UserModel.cs:13]: The difference with 'UserDetailsModel' is that in this class we don't have the list of accessControls (commented for the moment as not yet implemented).

File: src/IoTHub.Portal.Shared/Models/v1.0/UserModel.cs (hunk 6 above)
Comment (TLeoDev) [reply src/IoTHub.Portal.Shared/Models/v1.0/UserModel.cs:12]: I will do it in the next migration

File: src/IoTHub.Portal.Server/Controllers/v1.0/RolesController.cs [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+            {
+                return NotFound($"Role with name {currentRoleName} not found.");
+            }
+
+            // Mise à jour des propriétés du rôle
+            existingRole.Name = roleDetails.Name;
+            existingRole.Description = roleDetails.Description;
+
+            // Mise à jour des actions
+            existingRole.Actions = UpdateRoleActions(existingRole.Actions, roleDetails.Actions.Select(a => new Action { Name = a }).ToList());
```
[the remark this answers is not in the document]
Comment (TLeoDev) [reply src/IoTHub.Portal.Server/Controllers/v1.0/RolesController.cs]: outdated (old version before changes)

File: src/IoTHub.Portal.Application/Services/RoleManagementService.cs [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+            return roleEntity;
+        }
+
+        public async Task<Role?> UpdateRole(Role role)
+        {
+            ArgumentNullException.ThrowIfNull(role, nameof(role));
+            var roleEntity = await roleRepository.GetByIdAsync(role.Id);
+            if (roleEntity == null) return null;
+
+            // Mise à jour des propriétés
```
[the remark this answers is not in the document]
Comment (TLeoDev) [reply src/IoTHub.Portal.Application/Services/RoleManagementService.cs]: outdated

File: src/IoTHub.Portal.Infrastructure/Services/RRAC/RoleManagementService.cs
[the remark this answers is not in the document]
Comment (TLeoDev) [reply src/IoTHub.Portal.Infrastructure/Services/RRAC/RoleManagementService.cs:1]: I made this changes

File: src/IoTHub.Portal.Shared/Models/v1.0/UserModel.cs (hunk 6 above)
Comment (TLeoDev) [reply src/IoTHub.Portal.Shared/Models/v1.0/UserModel.cs]: I made theses changes

File: src/IoTHub.Portal.Shared/Models/v1.0/RoleDetailsModel.cs (hunk 1 above)
Comment (TLeoDev) [reply src/IoTHub.Portal.Shared/Models/v1.0/RoleDetailsModel.cs]: changed

File: src/IoTHub.Portal.Shared/Models/v1.0/UserModel.cs (hunk 4 above)
Comment (TLeoDev) [reply src/IoTHub.Portal.Shared/Models/v1.0/UserModel.cs]: Outdated (before the architecture change)

File: src/IoTHub.Portal.Shared/Models/v1.0/ActionModel.cs (hunk 2 above)
Comment (TLeoDev) [reply src/IoTHub.Portal.Shared/Models/v1.0/ActionModel.cs]: Yes, this class does'nt exist anymore with the new architecture

File: src/IoTHub.Portal.Domain/Entities/indexerAttribute.cs [hunk 9]
```diff
@@ -0,0 +1,11 @@
+﻿// Copyright (c) CGI France. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace IoTHub.Portal.Domain.Entities
+{
+    using System;
+
+    internal class indexerAttribute : Attribute
```
[the remark this answers is not in the document]
Comment (TLeoDev) [reply src/IoTHub.Portal.Domain/Entities/indexerAttribute.cs:8]: Done

File: src/IoTHub.Portal.Infrastructure/PortalDbContext.cs [hunk 10]
```diff
@@ -62,30 +63,26 @@ protected override void OnModelCreating(ModelBuilder modelBuilder)
...
+                .WithMany()
+                .HasForeignKey(ac => ac.RoleId)
+                .OnDelete(DeleteBehavior.Restrict);
 
             _ = modelBuilder.Entity<User>()
-               .HasMany(a => a.AccessControls);
-
-            _ = modelBuilder.Entity<Role>()
-                .HasMany(a => a.Actions);
+                .HasIndex(u => u.Email)
```
[the remark this answers is not in the document]
Comment (TLeoDev) [reply src/IoTHub.Portal.Infrastructure/PortalDbContext.cs:77]: It is not on ‘_ = modelBuilder.Entity<Role>()’ but on ‘_ = modelBuilder.Entity<User>()’: ‘_ = modelBuilder.Entity<User>()’, this index is used when a user does a filtered search (they can type in a keyword, which is compared with the name and email to see if it matches).

File: src/IoTHub.Portal.Server/Controllers/v1.0/DashboardController.cs [hunk 11]
```diff
@@ -3,7 +3,7 @@
 
 namespace IoTHub.Portal.Server.Controllers.v1._0
 {
-    using IoTHub.Portal.Shared.Security;
+    //using IoTHub.Portal.Shared.Security;
```
[the remark this answers is not in the document]
Comment (TLeoDev) [reply src/IoTHub.Portal.Server/Controllers/v1.0/DashboardController.cs:6]: Done

File: src/IoTHub.Portal.Server/Controllers/v1.0/DashboardController.cs [hunk 12]
```diff
@@ -23,7 +23,7 @@ public DashboardController(PortalMetric portalMetric)
         }
 
         [HttpGet("metrics", Name = "Get Portal Metrics")]
-        [Authorize(Policy = Policies.GetPortalMetrics)]
+        //[Authorize(Policy = Policies.GetPortalMetrics)]
```
[the remark this answers is not in the document]
Comment (TLeoDev) [reply src/IoTHub.Portal.Server/Controllers/v1.0/DashboardController.cs:26]: done

File: src/IoTHub.Portal.Server/Controllers/v1.0/SettingsController.cs [hunk 13]
```diff
@@ -13,7 +13,7 @@ namespace IoTHub.Portal.Server.Controllers.V10
     using Microsoft.AspNetCore.Mvc;
     using Microsoft.Extensions.Options;
     using IoTHub.Portal.Domain;
-    using IoTHub.Portal.Shared.Security;
+    //using IoTHub.Portal.Shared.Security;
```
[the remark this answers is not in the document]
Comment (TLeoDev) [reply src/IoTHub.Portal.Server/Controllers/v1.0/SettingsController.cs:16]: done

File: src/IoTHub.Portal.Server/Controllers/v1.0/GroupsController.cs [hunk 14]
```diff
@@ -0,0 +1,193 @@
...
+        /// <returns></returns>
+        [HttpDelete("{id}")]
+        [ProducesResponseType(StatusCodes.Status200OK)]
+        [ProducesResponseType(StatusCodes.Status404NotFound)]
+        public async Task<IActionResult> DeleteGroup(string id)
+        {
+            return Ok(await groupManagementService.DeleteGroup(id));
+            try
+            {
+                var result =await groupManagementService.DeleteGroup(id);
```
Comment (TLeoDev) [inline src/IoTHub.Portal.Server/Controllers/v1.0/GroupsController.cs:156]: The code in the try is never reached

File: src/IoTHub.Portal.Tests.Unit/Server/Controllers/v1.0/AccessControlsControllerTests.cs [hunk 15]
```diff
@@ -0,0 +1,268 @@
...
+            var result = await accessControlsController.EditAccessControlAsync(accessControlId, accessControl);
+
+            // Assert
+            Assert.IsNotNull(result);
+
+            this.mockRepository.VerifyAll();
+        }
+
+        /*[Test]
+        public async Task DeleteAccessControlShouldReturnExpectedBehavior()
```
[the remark this answers is not in the document]
Comment (TLeoDev) [reply src/IoTHub.Portal.Tests.Unit/Server/Controllers/v1.0/AccessControlsControllerTests.cs:234]: I made the changes, its working now

File: src/IoTHub.Portal.Application/Services/UserService.cs [hunk 16]
```diff
@@ -0,0 +1,130 @@
...
+            return toModel;
+
+        }
+
+        public async Task<bool> DeleteUser(string id)
+        {
+            var user = await userRepository.GetByIdAsync(id);
+            if (user is null)
+            {
+                throw new ResourceNotFoundException($"The User with the id {id} that you want to delete does'nt exist !");
```
[the remark this answers is not in the document]
Comment (TLeoDev) [reply src/IoTHub.Portal.Application/Services/UserService.cs:121]: Fixed

File: src/IoTHub.Portal.Application/Services/UserService.cs [hunk 17]
```diff
@@ -0,0 +1,130 @@
...
+        }
+
+        public async Task<UserDetailsModel?> UpdateUser(string id, UserDetailsModel user)
+        {
+            var userEntity = await this.userRepository.GetByIdAsync(id);
+            if (userEntity == null) throw new ResourceNotFoundException($"The User with the id {id} does'nt exist !");
+            var existingName = await this.userRepository.GetByNameAsync(user.Name);
+            if (existingName is not null)
+            {
+                throw new ResourceAlreadyExistsException($"The User tis the name {user.Name} already exist !");
```
[the remark this answers is not in the document]
Comment (TLeoDev) [reply src/IoTHub.Portal.Application/Services/UserService.cs:99]: Fixed

File: src/IoTHub.Portal.Application/Services/GroupService.cs [hunk 18]
```diff
@@ -0,0 +1,166 @@
...
+            };
+            return new PaginatedResult<GroupModel>(paginatedGroupDto.Data, paginatedGroupDto.TotalCount);
+        }
+
+        public async Task<GroupDetailsModel> CreateGroupAsync(GroupDetailsModel group)
+        {
+            var existingName = await this.groupRepository.GetByNameAsync(group.Name);
+            if (existingName is not null)
+            {
+                throw new ResourceAlreadyExistsException($"The Group tis the name {group.Name} already exist !");
```
[the remark this answers is not in the document]
Comment (TLeoDev) [reply src/IoTHub.Portal.Application/Services/GroupService.cs:83]: Fixed

**Comment by Metal-Mighty** [comment]
Use AutoFixtures in all Unit Tests, some are missing

**Review by Metal-Mighty — CHANGES_REQUESTED** [review]
Use AutoFixtures in all Unit Tests, some are missing
Add tests of limits in methods

**Review by Metal-Mighty — CHANGES_REQUESTED** [review]

**Pull request merged**